Introduce Span input and output support

iCharlesHu opened the pull request:
This PR adds support for `Span` as Input and Output:
- An overload of `run` that accepts `Span<Element: BitwiseCopyable>`
- An `SpanOutput` that vends `RawSpan` as output.

iCharlesHu committed:
Introduce Span input and output support

iCharlesHu committed:
WIP: Use RawSpan in OutputProtocol

iCharlesHu committed:
Adopting Span

glessard committed:
fix lifetime handling

glessard committed:
work around a SIL bug

iCharlesHu committed:
Use Mutex instead of LockedState

iCharlesHu committed:
Rename module name to Subprocess

iCharlesHu committed:
Drop struct Subprocess namespace

iCharlesHu committed:
Update proposal to drop struct Subprocess namespace

parkera commented on README.md:
```diff
@@ -160,14 +162,14 @@ While the Swift script above is functionally equivalent to the shell script, it
-import FoundationEssentials
+import Subprocess
 
 let gitResult = try await Subprocess.run(   // <- 0
```
```suggestion
let gitResult = try await run(   // <- 0
```

iCharlesHu replied:
On in testing I found you can't actually omit the `Subprocess` here otherwise you get
```
Use of 'run' refers to instance method rather than global function 'run(_:arguments:environment:workingDirectory:platformOptions:input:output:error:)' in module 'Subprocess'
```

iCharlesHu replied:
^^ That only existed under `XCTestCase` because of the name `run()`.

parkera commented on README.md:
```diff
@@ -195,123 +197,284 @@ Let's break down the example above:
+### The `run()` Family of Methods
 
-We propose a new `struct Subprocess`. Developers primarily interact with `Subprocess` via the static `run` methods which asynchronously executes a subprocess.
+We propose several `run()` functions that allows developers to asynchronously executes a subprocess.
```
```suggestion
We propose several `run()` functions that allow developers to asynchronously execute a subprocess.
```

parkera commented on README.md:
```diff
@@ -195,123 +197,284 @@ Let's break down the example above:
+///   - error: The method to use for redirecting the standard error.
+///   - body: The custom execution body to manually control the running process
+/// - Returns a CollectedResult containing the result of the run.
+@available(macOS 9999, *)
```
What kind of availability do we really need here? I assume it's 9999 because we don't have a macOS with `Span` yet?

iCharlesHu replied:
Remove unnecessary ones and just comment on the proposal that's it's Span.

parkera commented on README.md:
```diff
@@ -372,227 +535,198 @@ let result = try await Subprocess.run(
+/// executed. You can use this object to send signals to the
+/// child process as well as stream its output and error.
+@available(macOS 9999, *)
+public struct Execution<
```
If this is only valid in the scope of the closure, does it make sense to mark it as `~Escapable`?

parkera commented on README.md:
```diff
@@ -372,227 +535,198 @@ let result = try await Subprocess.run(
-        /// The process identifier of the current execution
-        public let processIdentifier: ProcessIdentifier
-    }
+/// An object that repersents a subprocess that has been
```
```suggestion
/// An object that represents a subprocess that has been
```

parkera commented on README.md:
```diff
@@ -372,227 +535,198 @@ let result = try await Subprocess.run(
     /// Accessing this property will **fatalError** if
     /// - `.output` wasn't set to `.redirectToSequence` when the subprocess was spawned;
     /// - This property was accessed multiple times. Subprocess communicates with
     ///   parent process via pipe under the hood and each pipe can only be consumed ones.
```
```suggestion
    ///   parent process via pipe under the hood and each pipe can only be consumed once.
```
(this typo is in a couple of places)

parkera commented on README.md:
```diff
@@ -372,227 +535,198 @@ let result = try await Subprocess.run(
-    }
+#if canImport(WinSDK)
+/// A platform independent identifier for a subprocess.
+@available(macOS 9999, *)
```
What does this availability mean for Windows?

parkera commented on README.md:
```diff
@@ -606,27 +740,26 @@ extension Subprocess.Execution {
+/// It consists of a signal to send to the child process and the
+/// duration allowed for the child process to exit before proceeding
+/// to the next step.
+@available(FoundationPreview 0.4, *)
```
I think this availability should be updated to match the rest, whatever we decide.

parkera commented on README.md:
```diff
@@ -606,27 +740,26 @@ extension Subprocess.Execution {
+@available(iOS, unavailable)
+@available(tvOS, unavailable)
+@available(watchOS, unavailable)
+public struct TeardownStep: Sendable, Hashable {
```
It's not immediately clear to me what the name of this type is really meant for. It's not possible to instantiate it, right? Should this function be static on `Signal` instead?

parkera commented on README.md:
```diff
@@ -678,147 +811,141 @@ extension Subprocess.Execution {
+        executing executable: Executable,
+        arguments: Arguments = [],
+        environment: Environment = .inherit,
+        workingDirectory: FilePath? = nil,
```
Should the ivar also be `FilePath?` instead of non-nullable?

parkera commented on README.md:
```diff
@@ -678,147 +811,141 @@ extension Subprocess.Execution {
+public final actor StandardInputWriter {
+    /// Write a sequence of UInt8 to the standard input of the subprocess.
+    /// - Parameter sequence: The sequence of bytes to write.
+    public func write<SendableSequence: Sequence<UInt8> & Sendable>(
```
I'm still unsure about this - do we really want to support *any* sequence of `UInt8` or do we want to require a `Collection` instead so we know we can do a bulk operation? Even if we want to allow it, shouldn't we provide concrete overrides of this that take a `Collection` or some other known-contiguous types like `[UInt8]` to avoid the overhead of casting in the implementation?

iCharlesHu replied:
Explicit `ArrayInput` and `DataInput`.

parkera commented on README.md:
```diff
@@ -678,147 +811,141 @@ extension Subprocess.Execution {
+    ) async throws
+    /// Write a String to the standard input of the subprocess.
+    /// - Parameter sequence: The string write.
+    public func write(
```
Does this deserve a defaulted-to-`.utf8` encoding argument?

parkera commented on README.md:
```diff
@@ -849,183 +976,179 @@ platformOptions.preSpawnProcessConfigurator = { _, fileAttr in
+    /// process in case the parent task is cancelled before
+    /// the child proces terminates.
+    /// Always ends in sending a `.kill` signal at the end.
+    public var teardownSequence: [TeardownStep] = []
```
I guess this is where these are used, but how do you create them other than the one static func?

parkera commented on README.md:
```diff
@@ -849,183 +976,179 @@ platformOptions.preSpawnProcessConfigurator = { _, fileAttr in
+        // of the user account to run as.
+        public var username: String
+        // The clear-text password for the account.
+        public var password: String
```
I assume this mirrors an underlying Win32 API?

iCharlesHu replied:
Yes. As bad as it is the Win32 API literately says "plain text password"... 😅

parkera commented on README.md:
```diff
@@ -1059,93 +1184,99 @@ _(For more information on these values, checkout Microsoft's documentation [here
+
+/// A concrete `Input` type for subprocesses that reads input
+/// from a given sequence of `UInt8` (such as `Data` or `[UInt8]`).
+public struct UInt8SequenceInput<
```
I have the same question here as above - do we really want to provide this? Because accidentally falling back to a byte-by-byte write for each element of some collection could end up being very expensive, and surprising.

parkera commented on README.md:
```diff
@@ -1155,130 +1286,136 @@ extension Subprocess.InputProtocol where Self == Subprocess.FileDescriptorInput
+### `OutputProtocol`
+
 
+`OutputProtocol` defines the set of methods that a type must implement to serve as the output target for a subprocess. Similarly to `InputProtocol`, developers should utilize the built-in concrete `Output` types provided with `Subprocess` whenever feasible; alternatively, they can create a type that conforms to `PipeBasedOutputProtocol` and implements `func output(from:) throws -> Data` for custo ...
```
```suggestion
`OutputProtocol` defines the set of methods that a type must implement to serve as the output target for a subprocess. Similarly to `InputProtocol`, developers should utilize the built-in concrete `Output` types provided with `Subprocess` whenever feasible; alternatively, they can create a type that conforms to `PipeBasedOutputProtocol` and implements `func output(from:) throws -> OutputType` for custom outputs.
```

parkera commented on README.md:
```diff
@@ -1300,47 +1437,63 @@ extension Subprocess.OutputProtocol where Self == Subprocess.FileDescriptorOutpu
-    public static func collectString(
-        upTo limit: Int = 128 * 1024,
+    public static func string(
+        limit: Int,
```
It would be good to specify that this value is measured in bytes.

parkera commented on README.md:
```diff
@@ -1349,162 +1502,159 @@ print("curl output: \(curl.standardOutput?.count)")
+    /// The result returned by the closure passed to `.run` methods
+    public let value: Result
+
+    internal init(terminationStatus: TerminationStatus, value: Result) {
```
Is this public, or should we omit from the pitch?

parkera commented on README.md:
```diff
@@ -1349,162 +1502,159 @@ print("curl output: \(curl.standardOutput?.count)")
+        _ newValue: [String : String]
+    ) -> Self
 
 #if !os(Windows)
```
Is Win32 incompatible with `Data`, requiring `String` values only?

iCharlesHu committed:
WIP: rename PipeBasedInputProtocol

iCharlesHu committed:
Move output(from:) to PipeOutputProtocol and rename files to better suited as a package

iCharlesHu committed:
Rename Pipe{Input|Output}Protocol to Managed{Input|Output}Protocol

iCharlesHu committed:
Clean up @availability markings. Now only the CollectedResult run family requires Span availability

itingliu commented on README.md:
```diff
@@ -189,168 +194,320 @@ Let's break down the example above:
+///     of the closure.
+public func run<Result, Output: OutputProtocol, Error: OutputProtocol>(
+    _ configuration: Configuration,
+    isolation: isolated (any Actor)? = #isolation,
```
CAn we make sure the order of the arguments stay consistent? Previously the arguments are in this order: 
- input
- output
- error
- isolation
- body

But here isolation is moved up to before output and error. Same with the one below

iCharlesHu replied:
Ahh good catch! Thanks!

itingliu commented on README.md:
```diff
@@ -1041,470 +1128,430 @@ platformOptions.preSpawnProcessConfigurator = { flag, startupInfo in
+    /// Close the FileDescriptor for writing
+    func closeWriteFileDescriptor() throws
+
+    /// Capture the output from the subprocess up to maxSize
```
what is maxSize?

iCharlesHu committed:
Fix Linux tests

iCharlesHu committed:
Fix Windows build

iCharlesHu committed:
Keep run parameter order consistent

iCharlesHu committed:
Remove 'default argument' run() overloads

iCharlesHu committed:
Separate Subprocess into core module and SubprocessFoundation module [Part 1]

Core module will not depend on Foundation while SubprocessFoundation will depend on Foundation and extend Subprocess to work with Foundation types

iCharlesHu committed:
Create Buffer and move tests to Swift Testing

iCharlesHu committed:
Added description to SubprocessError

iCharlesHu committed:
Rename _CShims to _SubprocessCShims

iCharlesHu committed:
Add BufferOutput

iCharlesHu committed:
Added testStringInput

iCharlesHu committed:
Updated proposal to reflex the new SubprocessFoundation module

iCharlesHu committed:
Added SubprocessError to the proposal

iCharlesHu committed:
Make UnderlyingError a concrete type itself
